dwave/system/temperatures.py

```diff
@@ -0,0 +1,488 @@
+    where nu is the energy lost in flipping spin i against its current 
+    assignment (the effective field). 
+    
+    The problem is a convex root solving problem, and is solved with scipy 
+    optimize.
+
+    If the distribution is not Boltzmann with respect to the BQM provided, as
+    may be the case for heuristic samplers (such as annealers). The temperature
+    estimate can be interpretted as characterizing only a rate of local 
+    excitations. In the case of sample sets obtained from D-Wave quantum 
+    annealing the temperature can be identified with a physical temperature via
+    a late-anneal freeze-out phenomena.
+    
+    Args:
+        bqm (:obj:`dimod.BinaryQuadraticModel`, optional):
+            Binary quadratic model describing sample distribution.
+            If bqm and site_energy are both None, then by default 
+            100 samples are drawn from the default DWaveSampler(), 
+            with bqm defaulted as described.
+        sampleset (:class:`dimod.SampleSet`, optional):
+            A set of samples, assumed to be fairly sampled from
+            a Boltzmann distribution characterized by bqm.
+        site_energy (2d numpy array, optional):
+            The effective fields associated to a sampleset and bqm. Rows
+            denote spins, columns samples.
+            An effective field matrix derived from the 
+            bqm and sampleset if not provided.
+        bootstrap_size (int, optional, default=0):
+            Number of bootstrap estimators to calculate.
+        seed (int, optional)
+            Seeds the bootstap method (if provided) allowing reproducibility
+            of the estimators given bqm and samples_like.
+        Tguess (float, optional):
+            Seeding the optimize process can enable faster convergence.
+
+    Returns:
+        tuple of float and numpy array:
+            (T_estimate,T_bootstrap_estimates)
+
+            *T_estimate*: a temperature estimate
+            *T_bootstrap_estimates*: a numpy array of bootstrap estimators
+
+    Examples:
+       Draw samples from the default DWaveSampler() for a large spin-glass 
+       problem (random couplers J, zero external field h).
+       Establish a temperature estimate by maximum pseudo-likelihood. 
+
+       Note that due to the complicated freeze-out properties of hard models,
+       such as large scale spin-glasses, deviation from a classical Boltzmann 
+       distribution is anticipated.
+       Nevertheless, the T estimate can always be interpretted as an estimator
+       of local excitations rates. For example T will be 0 if only 
+       local minima are returned (even if some of the local minima are not 
+       ground states).
+       
+       >>> import dimod
+       >>> from dwave.system.temperatures import maximum_pseudolikelihood_temperature
+       >>> from dwave.system.samplers import DWaveSampler
+       >>> from random import random
+       >>> sampler = DWaveSampler()     # doctest: +SKIP
```

Comment:
why skip these lines? sampler will be set to a mock DWaveSampler, a BQM will be created

Reply:
Okay - thanks for that clarification. In the past I've had errors thrown when using DWaveSampler(), perhaps it is because the mock sampler lacked all the structure (attributes and properties) of a QPU sampler, that shouldn't be an issue in these examples.